File: sections/13_offline_processing/problems/b_range_count_at_most/solution.py
"""Reference solution for B. Range Count At Most."""

import sys


class Fenwick:
    def __init__(self, n: int) -> None:
        self.bit = [0] * (n + 1)

    def add(self, idx: int, delta: int) -> None:
        while idx < len(self.bit):
            self.bit[idx] += delta
            idx += idx & -idx

    def sum(self, idx: int) -> int:
        total = 0
        while idx > 0:
            total += self.bit[idx]
            idx -= idx & -idx
        return total

    def range_sum(self, left: int, right: int) -> int:
        return self.sum(right) - self.sum(left - 1)


def solve_case(a: list[int], queries: list[tuple[int, int, int]]) -> list[int]:
    values = sorted((value, pos) for pos, value in enumerate(a, start=1))
    ordered = sorted((x, l, r, idx) for idx, (l, r, x) in enumerate(queries))
    bit = Fenwick(len(a))
    ans = [0] * len(queries)
    ptr = 0
    for x, l, r, idx in ordered:
        while ptr < len(values) and values[ptr][0] <= x:
            _, pos = values[ptr]
            bit.add(pos, 1)
            ptr += 1
        ans[idx] = bit.range_sum(l, r)
    return ans


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        q = data[idx + 1]
        idx += 2
        a = data[idx:idx + n]
        idx += n
        queries: list[tuple[int, int, int]] = []
        for _ in range(q):
            l = data[idx]
            r = data[idx + 1]
            x = data[idx + 2]
            idx += 3
            queries.append((l, r, x))
        out.extend(map(str, solve_case(a, queries)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/13_offline_processing/problems/b_range_count_at_most/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
    vector<int> bit;

    explicit Fenwick(int n) : bit(n + 1, 0) {}

    void add(int idx, int delta) {
        while (idx < (int)bit.size()) {
            bit[idx] += delta;
            idx += idx & -idx;
        }
    }

    int sum(int idx) const {
        int total = 0;
        while (idx > 0) {
            total += bit[idx];
            idx -= idx & -idx;
        }
        return total;
    }

    int range_sum(int left, int right) const {
        return sum(right) - sum(left - 1);
    }
};

struct Query {
    int l;
    int r;
    long long x;
    int idx;
};

vector<int> solve_case(const vector<long long> &a, vector<Query> queries) {
    vector<pair<long long, int>> values;
    values.reserve(a.size());
    for (int i = 0; i < (int)a.size(); ++i) {
        values.push_back({a[i], i + 1});
    }
    sort(values.begin(), values.end());
    sort(queries.begin(), queries.end(), [](const Query &lhs, const Query &rhs) {
        return lhs.x < rhs.x;
    });

    Fenwick bit((int)a.size());
    vector<int> ans(queries.size());
    int ptr = 0;
    for (const Query &query : queries) {
        while (ptr < (int)values.size() && values[ptr].first <= query.x) {
            bit.add(values[ptr].second, 1);
            ++ptr;
        }
        ans[query.idx] = bit.range_sum(query.l, query.r);
    }
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n, q;
        cin >> n >> q;
        vector<long long> a(n);
        for (long long &x : a) cin >> x;
        vector<Query> queries;
        queries.reserve(q);
        for (int i = 0; i < q; ++i) {
            Query query;
            cin >> query.l >> query.r >> query.x;
            query.idx = i;
            queries.push_back(query);
        }
        vector<int> ans = solve_case(a, queries);
        for (int value : ans) {
            cout << value << '\n';
        }
    }

    return 0;
}

